nixos/modules/programs/gnupg.nix

```diff
@@ -95,29 +95,101 @@ in
   config = mkIf cfg.agent.enable {
     # This overrides the systemd user unit shipped with the gnupg package
     systemd.user.services.gpg-agent = mkIf (cfg.agent.pinentryFlavor != null) {
-      serviceConfig.ExecStart = [ "" ''
-        ${cfg.package}/bin/gpg-agent --supervised \
-          --pinentry-program ${pkgs.pinentry.${cfg.agent.pinentryFlavor}}/bin/pinentry
-      '' ];
+      unitConfig = {
+        Description = "GnuPG cryptographic agent and passphrase cache";
+        Documentation = "man:gpg-agent(1)";
+        Requires = [ "gpg-agent.socket" ];
+      };
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/gpg-agent --supervised \
+            --pinentry-program ${pkgs.pinentry.${cfg.agent.pinentryFlavor}}/bin/pinentry
```

Comment:
I would really avoid this now that we can: setting the pinentry program for the systemd service only has been a constant source of pain for a number of users. See all the issues we have about pinentry[1][2][3][4].
Upstream also said `--pinentry-program` is only meant for testing: you should set the path from the gnupg configuration file

If you want to preserve the systemd units, that ok, but at least make an option to enable them.

[1]: https://discourse.nixos.org/t/cant-make-gpg-agent-work-with-ssh-on-a-non-gui-system/24750
[2]: https://discourse.nixos.org/t/cant-get-gnupg-to-work-no-pinentry/15373
[3]: https://github.com/NixOS/nixpkgs/issues/35464
[4]: https://github.com/NixOS/nixpkgs/issues/210375

Reply:
I'm all for removing the systemd units.  My only worry is: will we be breaking anyone by removing the socket activation?  Like maybe ssh-agent compatibility?

It may still be a good idea regardless, but we'd probably need some documentation.

Reply:
We should probably add a changelog entry either way because this is most likely going to cause regressions and they are hard to debug with gnupg shit error messages ala doesn't work.

Reply:
I've moved this to /etc/